Lib/abc.py

```diff
@@ -121,6 +121,49 @@ def _abc_caches_clear(cls):
             """Clear the caches (for debugging or testing)."""
             _reset_caches(cls)
 
+def update_abstractmethods(cls):
+    """Repopulate the abstract methods of an abstract class, or a subclass of
+    an abstract class.
+
+    If a class has had one of its abstract methods implemented after the
+    class was created, the method will not be considered implemented until
+    this function is called. Alternatively, if a new abstract method has been
+    added to the class, it will only be considered an abstract method of the
+    class after this function is called.
+
+    This function should be called before any use is made of the class,
+    usually in class decorators that add methods to the subject class.
+
+    Returns cls, to allow usage as a class decorator.
+
+    If cls is has any subclasses, raises a RuntimeError.
+
+    If cls is not an instance of ABCMeta, does nothing.
+    """
+    if not hasattr(cls, '__abstractmethods__'):
+        # We check for __abstractmethods__ here because cls might by a C
+        # implementation or a python implementation (especially during
+        # testing), and we want to handle both cases.
+        return cls
+
+    if cls.__subclasses__():
+        raise RuntimeError("cannot update abstract methods of class after"
+                           " subclassing")
```

Comment:
I'm not sure I like this. This function is potentially expensive. It could even be detrimental -- I could imagine some very dynamic code that's messing around dynamically with a base class and then calls `update_abstractmethods()` for all subclasses. So I propose to remove this check (and the documentation about it).

Reply:
That might be okay, However, removing this check means that we have to call `update_abstractmethods` on all of `cls`'s subclasses, meaning that `cls.__subclasses__()` will still be called, so performance will not improve.

Reply:
Only if you're not sure that there are no subclasses. In `@dataclass` we can assume there are no subclasses, and that's so in most class decorators (which are only applicable to freshly created classes).

I recommend that `update_abstractmethods()` should *not* try to fix up subclasses -- it's just the case that an ambitious dynamic class hacker could take two documented features (this and `__subclasses__()`) and create something that automatically updates affected subclasses.

Reply:
I think that could cause a problem if user chain two decorators: one that updates subclasses and one that doesn't

```python

def decoratorA(cls):
  """implements methods and creates a subclass, calling
  update_abstractmethods on cls and its subclasses
  """
  ...

def decoratorB(cls):
  """implements methods only, not updating subclasses
  """
  ...

@decoratorB
@decoratorA
class A(SomeABC):
  ...

# A's subclass is now out of sync with decoratorB's changes
```

Reply:
If you do that, you get what you deserve.

Please keep it simple.

Reply:
It's implemented in the latest commit, but I must say I'm not sure what you mean. Who "gets what they deserve"?
* Whoever implemented `decoratorA` (which may well be `@dataclass`) was right to assume there are no subclasses, since it did not create them, and wanted to avoid unnecessary checks.
* Whoever implemented `decoratorB` did their due diligence by making sure it returned the class and its subclasses in a synced state.
* Whoever wrote `class A` may be totally unaware that `decoratorB` creates subclasses or that `decoratorA` doesn't take them into account. Even if they did,  they would need to implement a fix themselves, which would require at the least to iterate through `A.__subclasses__()`. Even that won't keep them 100% safe, however, since these subclasses can have subclasses of their own. Very far fetched, perhaps, but that's exactly what very dynamic code might produce.

Reply:
Truth be told, if we do assume that decorators and tools are allowed to subclass (and perhaps other permanent operations like instancing) before calling `update_abstractmethods`, I'm leaning closer and closer to automatic solution as proposed by Bar and Reymond.